zip-0315.rst

```diff
@@ -1,9 +1,629 @@
+they become spendable than trusted TXOs. This provides an improved trade-off
+between latency on the one hand, and reliability and safety on the other.
+
+
+Specification
+=============
+
+Long-term storage of funds
+--------------------------
+
+It is RECOMMENDED that wallets only hold funds as shielded in the long term;
+that is, if a wallet supports receiving transparent funds (or supports
+importing a seed from another wallet that might have done so), then it SHOULD
+either auto-shield or opportunistically shield such funds by default.
+
+Following this recommendation improves both collective privacy and the user's
+individual privacy, by maximizing the size of the note anonymity set over time.
+
+The remainder of this specification assumes a wallet that follows the above
+recommendation, except where explicitly noted.
+
+A wallet MAY allow users to disable auto-shielding.
+
+
+Trusted and untrusted TXOs
+--------------------------
+
+A wallet SHOULD treat received TXOs that are outputs of transactions created
+by the same wallet, as trusted TXOs. Wallets MAY enable users to mark specific
+external transactions as trusted, allowing their received TXOs also to be
+classified as trusted TXOs.
+
+A wallet SHOULD have a policy that is clearly communicated to the user for
+the number of confirmations needed to spend untrusted and trusted TXOs
+respectively. The following confirmation policy is RECOMMENDED:
+
+* 10 confirmations, for untrusted TXOs;
+* 3 confirmations, for trusted TXOs.
+
+Rationale for the given numbers of confirmations
+''''''''''''''''''''''''''''''''''''''''''''''''
+
+The rationale for choosing three confirmations for trusted TXOs is that
+empirically, reorgs are usually less than three blocks.
+
+The consequences of attempting to spend a trusted TXO may be less severe in the
+case of a rollback than the consequences of attempting to spend an untrusted TXO.
+The value received from a trusted TXO should always be recoverable, whereas
+recovering value received from an untrusted TXO may require the user to request
+that funds are re-sent.
+
+
+Categories of TXOs according to spendability
+--------------------------------------------
+
+A TXO is *spendable*, relative to a given block chain and wallet state,
+if and only if all of the following are true in that state:
+
+* the TXO is unspent;
+* the TXO is not committed to be spent in another transaction created
+  by this wallet; and
+* the wallet has the TXO's spending key (for whatever protocol the TXO uses).
+
+A TXO is *confirmed-spendable*, relative to a given block chain and
+wallet state, if and only if all of the following are true in that state:
+
+* the wallet is synchronized; and
+* the TXO is spendable; and
+* either auto-shielding is disabled or the TXO is shielded; and
+* the TXO is trusted and has at least the required confirmations for
+  trusted TXOs, or it is untrusted and has at least the required
+  confirmations for untrusted TXOs.
```

Comment:
There are several issues with this definition:
* Requiring the wallet to be synchronized is too strong. It's sufficient to replace that with a requirement that we know the TXO is unspent at the current chain tip. This in turn requires that we've synchronized the nullifier set to that tip, but that belongs in the definition of "known-spendable".
* It assumes that transparent TXOs should be treated as confirmed-spendable if auto-shielding is disabled (and other conditions are met). That's not correct because the alternative to auto-shielding, now that we've decided that opportunistic shielding should be excluded for privacy reasons, is manually triggered shielding. In that case the funds are still not confirmed user-spendable until the user has shielded them.

```suggestion
A TXO is *confirmed user-spendable*, relative to a given block chain and
wallet state, if and only if all of the following are true in that state:

* the TXO is known-spendable; and
* the TXO is shielded, or the wallet allows spending transparent
  TXOs in user-initiated transactions contrary to recommendation; and
* the TXO is trusted and has at least the required confirmations for
  trusted TXOs, or it is untrusted and has at least the required
  confirmations for untrusted TXOs.
```

Reply:
I disagree with this state being called "confirmed user-spendable", and with the change to the second bullet point. Note that the existing definition of confirmed-spendable was chosen to ensure that the "confirmed-spendable balance" includes only outputs that a user-initiated transaction can spend, and manual shielding is a user-initiated transaction. The fact that a wallet doesn't allow spending transparent TXOs directly to _an external address_ is a separate consideration.

The point of having the synchronized aspect be part of "confirmed-spendable" rather than (what is now called) "known-spendable" was to ensure that wallets don't report zero balance in the short time between starting up and catching up to a recent chain tip (instead they should show that balance as part of "pending"). Under the current definition, "unconfirmed-spendable" (the sum of which is the pending balance) omits any TXO for which the wallet's view is out-of-date, which will be all of them.